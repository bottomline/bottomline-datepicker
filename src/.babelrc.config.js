const { declare } = require('@babel/helper-plugin-utils');

function getEnvOption(envOptions, env, preset) {
  return envOptions[env] ? envOptions[env][preset] : {};
}

module.exports = declare((api, presetOptions = {}) => {
  // see docs about api at https://babeljs.io/docs/en/config-files#apicache
  api.assertVersion('^7.0.0');

  const { env: envOptions = {}, ...options } = presetOptions;

  return {
    presets: [
      ['@babel/preset-env', {
        modules: false,
        ...options['@babel/preset-env']
      }],
      '@babel/preset-react'
    ],
    plugins: [
      '@babel/plugin-proposal-class-properties',
      ['@babel/plugin-proposal-object-rest-spread', {
        useBuiltIns: true,
        ...options['@babel/plugin-proposal-object-rest-spread']
      }]
    ],
    env: {
      test: {
        presets: [
          ['@babel/preset-env', {
            targets: {
              node: 'current'
            },
            ...getEnvOption(envOptions, 'test', '@babel/preset-env')
          }],
          '@babel/preset-react'
        ],
        plugins: [
          '@babel/plugin-proposal-class-properties',
          ['@babel/plugin-proposal-object-rest-spread', {
            useBuiltIns: true,
            ...getEnvOption(envOptions, 'test', '@babel/plugin-proposal-object-rest-spread')
          }]
        ]
      }
    }
  };
});
