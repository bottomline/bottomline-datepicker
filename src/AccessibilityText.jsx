import injectSheet from 'react-jss';
import PropTypes from 'prop-types';
import React from 'react';

const styles = {
  root: {
    position: 'absolute',
    width: 1,
    height: 1,
    margin: -1,
    padding: 0,
    overflow: 'hidden',
    clip: 'rect(0, 0, 0, 0)',
    border: 0
  }
};

const AccessibilityText = ({ classes, children, ...rest }) => (
  <p className={classes.root} {...rest}>{children}</p>
);

AccessibilityText.propTypes = {
  classes: PropTypes.objectOf(PropTypes.string).isRequired,
  children: PropTypes.node.isRequired
};

export const AccessibilityTextBase = AccessibilityText;
export default injectSheet(styles)(AccessibilityText);
