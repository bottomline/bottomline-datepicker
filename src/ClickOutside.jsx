import React, { useEffect, useRef } from 'react';
import PropTypes from 'prop-types';

const ClickOutside = ({
  children, onClickOutside
}) => {
  const node = useRef();

  useEffect(() => {
    const handleKeyDown = e => {
      if (e.keyCode === 27) { // Escape
        onClickOutside();
      }
    };
    document.addEventListener('keydown', handleKeyDown);
    return () => {
      document.addEventListener('keydown', handleKeyDown);
    };
  }, [onClickOutside]);

  useEffect(() => {
    const handleClick = e => {
      if (e.target.contains(node.current)) onClickOutside();
    };
    window.addEventListener('resize', onClickOutside);
    document.addEventListener('click', handleClick);

    return () => {
      window.removeEventListener('resize', onClickOutside);
      document.removeEventListener('click', handleClick);
    };
  }, [onClickOutside]);

  return (
    <div ref={node}>{children}</div>
  );
};

ClickOutside.propTypes = {
  children: PropTypes.node.isRequired,
  inputEl: PropTypes.shape({}),
  onClickOutside: PropTypes.func.isRequired
};

ClickOutside.defaultProps = {
  inputEl: undefined
};

export default ClickOutside;
