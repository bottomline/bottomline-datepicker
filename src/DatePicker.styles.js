export default () => ({
  datepicker: {
    position: 'relative'
  }
});
