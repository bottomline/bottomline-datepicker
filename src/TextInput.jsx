import { Input } from '@glu/form-components';
import { CaretDownIcon, CaretUpIcon, CalendarIcon } from '@glu/icons-react';
import { AccessibilityText } from '@glu/utilities-react';
import { withStyles, withTheme } from '@glu/theming';
import PropTypes from 'prop-types';
import React from 'react';
import { stringToDates } from './utils/formatString.js';
import { isValidSelection } from './utils/dateCompare.js';
import styles from './TextInput.styles';

const TextInput = ({
  theme,
  classes,
  dateMax,
  dateMin,
  error,
  handleApply,
  htmlId,
  inputValue,
  labelText,
  name,
  open,
  placeholder,
  rangeDelimiter,
  screenReaderLabel,
  selectedEndDate,
  selectedStartDate,
  setHead,
  setInputValue,
  setOpen,
  setSelectedEndDate,
  setSelectedStartDate,
  singleDatePicker
}) => {
  const handleInputChange = (event) => {
    setInputValue(event.target.value);
    const [start, end] = stringToDates(event.target.value, rangeDelimiter);

    // Set start first
    if (start.isValid()) {
      if (isValidSelection(start, [dateMin, dateMax])) {
        setHead(start);
        setSelectedStartDate(start);
        // Finish workflow early if only choosing one date
        if (singleDatePicker) {
          setSelectedEndDate(start);
        }
        // Write end date
        if (end.isValid()) {
          if (isValidSelection(end, [dateMin, dateMax])
              && end.valueOf() >= start.valueOf()) {
            setSelectedEndDate(end);
          }
        }
      }
    }
  };

  const handleToggleClick = () => {
    setOpen(!open);
  };

  const handleInputKeyPress = (event) => {
    if (event.key === 'Enter') { // Enter
      if (open) handleApply(selectedStartDate, selectedEndDate);
      else setOpen(true);
    }
  };

  return (
    <>
      <Input
        autoComplete="off"
        className={classes.inputWrapper}
        labelText={labelText}
        screenReaderLabel={screenReaderLabel}
        data-qa="text-input"
        onChange={handleInputChange}
        onFocus={() => { setOpen(true); }}
        onKeyPress={handleInputKeyPress}
        type="text"
        value={inputValue}
        placeholder={placeholder}
        id={htmlId}
        name={name}
        error={error}
      />
      <button
        data-qa="toggle-button"
        id="toggle"
        className={`${classes.toggle} ${(!labelText || screenReaderLabel) ? classes.screenReaderLabel : ''}`}
        type="button"
        onClick={handleToggleClick}
      >
        <AccessibilityText>Toggle Dropdown</AccessibilityText>
        {theme.name === 'legacy'
          ? <CalendarIcon className={classes.inputIcon} />
          : (
            open
              ? <CaretUpIcon className={classes.inputIcon} />
              : <CaretDownIcon className={classes.inputIcon} />
          )
          }
      </button>
    </>
  );
};

TextInput.propTypes = {
  theme: PropTypes.shape({}),
  classes: PropTypes.shape({}).isRequired,
  dateMax: PropTypes.string.isRequired,
  dateMin: PropTypes.string.isRequired,
  error: PropTypes.string.isRequired,
  handleApply: PropTypes.func.isRequired,
  htmlId: PropTypes.string.isRequired,
  inputValue: PropTypes.string.isRequired,
  labelText: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  open: PropTypes.bool.isRequired,
  placeholder: PropTypes.string.isRequired,
  rangeDelimiter: PropTypes.string.isRequired,
  screenReaderLabel: PropTypes.bool.isRequired,
  selectedEndDate: PropTypes.shape({}),
  selectedStartDate: PropTypes.shape({}),
  setHead: PropTypes.func.isRequired,
  setInputValue: PropTypes.func.isRequired,
  setOpen: PropTypes.func.isRequired,
  setSelectedEndDate: PropTypes.func.isRequired,
  setSelectedStartDate: PropTypes.func.isRequired,
  singleDatePicker: PropTypes.bool.isRequired
};

TextInput.defaultProps = {
  theme: { name: 'ashTheme' },
  selectedStartDate: undefined,
  selectedEndDate: undefined
};

export default withTheme(withStyles(styles)(TextInput));
