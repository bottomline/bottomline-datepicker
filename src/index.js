export { default as DatePicker, BareDatePicker } from './DatePicker';

// eslint-disable-next-line camelcase
export { default as defaultStrings_en_US } from './locale/defaultStrings_en_US';
