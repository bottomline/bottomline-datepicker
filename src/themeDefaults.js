/* istanbul ignore file */
import { registerDefaultsForModule } from '@glu/theming';
import legacy from './themes/legacyTheme';
import ash from './themes/ashTheme';

const themes = { ash, legacy };

registerDefaultsForModule(
  'datepickerReact', ({ palette, typography, name }) => themes[name](palette, typography)
);
