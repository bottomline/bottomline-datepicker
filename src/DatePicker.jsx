import { withStyles, withTheme } from '@glu/theming';
import moment from 'moment';
import PropTypes from 'prop-types';
import React, { useState } from 'react';
import Dropdown from './Dropdown';
import TextInput from './TextInput';
import ClickOutside from './ClickOutside';
import { datesToString } from './utils/formatString.js';
import styles from './DatePicker.styles';
import './themeDefaults';

const DatePicker = ({
  classes,
  className,
  dataQa,
  dateEnd,
  dateStart,
  htmlId,
  name,
  onChangeCallback,
  screenReaderLabel,
  ...rest
}) => {
  const [open, setOpen] = useState(false);
  const [selectedStartDate, setSelectedStartDate] = useState(dateStart
    ? moment(dateStart)
    : undefined);
  const [selectedEndDate, setSelectedEndDate] = useState(dateEnd
    ? moment(dateEnd)
    : undefined);
  const [confirmedStartDate, setConfirmedStartDate] = useState(selectedStartDate);
  const [confirmedEndDate, setConfirmedEndDate] = useState(selectedEndDate);
  const [head, setHead] = useState(selectedStartDate ? moment(selectedStartDate) : moment());
  const [inputValue, setInputValue] = useState(
    datesToString(selectedStartDate, selectedEndDate, rest)
  );

  const handleApply = (start, end) => {
    setOpen(false);
    setConfirmedStartDate(start);
    setConfirmedEndDate(end);
    onChangeCallback({
      value: {
        date: moment(start),
        dateEnd: moment(end)
      }
    });
    setInputValue(datesToString(start, end, rest));
  };

  const handleClose = () => {
    setOpen(false);
    setSelectedStartDate(confirmedStartDate);
    setSelectedEndDate(confirmedEndDate);
  };

  return (
    <ClickOutside onClickOutside={handleClose}>
      <div data-qa={dataQa ? `${dataQa}` : `datepicker-${htmlId}`} className={`${classes.datepicker} ${className}`}>
        <TextInput {...{
          handleApply,
          handleClose,
          htmlId,
          inputValue,
          name,
          open,
          screenReaderLabel,
          selectedEndDate,
          selectedStartDate,
          setHead,
          setInputValue,
          setOpen,
          setSelectedEndDate,
          setSelectedStartDate,
          ...rest
        }}
        />
        {open && (

        <Dropdown {...{
          head,
          onApply: handleApply,
          onClose: handleClose,
          screenReaderLabel,
          selectedEndDate,
          selectedStartDate,
          setHead,
          setSelectedEndDate,
          setSelectedStartDate,
          ...rest
        }}
        />
        )}
      </div>
    </ClickOutside>
  );
};

DatePicker.propTypes = {
  /** Toggle whether to always show calendar when a range is selected. */
  alwaysShowCalendars: PropTypes.bool,

  /** Custom classes for the apply button */
  applyButtonClasses: PropTypes.string,

  /** Hide the apply and cancel buttons, and automatically apply a new selection */
  autoApply: PropTypes.bool,

  /** Custom classes for container that wraps action buttons */
  buttonClasses: PropTypes.string,

  /** Custom classes for the cancel button */
  cancelButtonClasses: PropTypes.string,

  /** Classes provided by the theme */
  classes: PropTypes.shape({}).isRequired,

  /** Class passed to parent element for extra decoration. */
  className: PropTypes.string,

  /** String given to parent elements data-qa html attribute for easy targeting */
  dataQa: PropTypes.string,

  /** The end of initial selected date range. Must be ISO format 'YYYY-MM-DD' */
  dateEnd: PropTypes.string,

  /** The latest date a user may select. Must be ISO format 'YYYY-MM-DD' */
  dateMax: PropTypes.string,

  /** The earliest date a user may select. Must be ISO format 'YYYY-MM-DD' */
  dateMin: PropTypes.string,

  /** The start of initial selected date range. Must be ISO format 'YYYY-MM-DD' */
  dateStart: PropTypes.string,

  /** Error message. will add styles to input if present. */
  error: PropTypes.string,

  /** Format to display date string in input */
  format: PropTypes.string,

  /** Function which is passed each visible day and returns string of classNames. */
  getClassNamesForDate: PropTypes.func,

  /** string assigned to HTML id attribute of component */
  htmlId: PropTypes.string,

  /* Function which is passed each visible date and returns boolean for if it is selectable. */
  isSelectableDate: PropTypes.func,

  /** Display weeks as ISO week numbers; weekNumbers must also be true. */
  isoWeekNumbers: PropTypes.bool,

  /** String used for label value */
  labelText: PropTypes.string,

  /** Toggle to control calendar navigation. Defaults to true. Can be set to
   * false to navigate calendars independently. */
  linkedCalendars: PropTypes.bool,

  /** Show select dropdown for the Month in the calendar header */
  monthDropdown: PropTypes.bool,

  /** string passed to html name attribute of text input */
  name: PropTypes.string,

  /** The number of calendars to display side by side. */
  numberOfMonths: PropTypes.number,

  /** Function that triggers once a date is selected or changed */
  onChangeCallback: PropTypes.func,

  /** String passed as html placeholder text for input. */
  placeholder: PropTypes.string,

  /**
  * Whether the picker appears below (default) or above the HTML element it's attached to.
  * ("down"/"up") */
  positionY: PropTypes.string,

  /**
   * Whether the picker appears aligned to the left, to the right, or centered
   * ("left"/"right"/"center") */
  positionX: PropTypes.string,

  /** Delimiter used to separate date string in text input */
  rangeDelimiter: PropTypes.string,

  /** Object with list of predefined ranges */
  rangePresets: PropTypes.shape({}),

  /** Toggle range preset feature */
  ranges: PropTypes.bool,

  /** Toggle to hide the label for all but screenreaders */
  screenReaderLabel: PropTypes.bool,

  /** Show "custom range" option to display calendar */
  showCustomRangeLabel: PropTypes.bool,

  /** Interacting with calendar sets a single date. */
  singleDatePicker: PropTypes.bool,

  /** Show timePicker */
  timePicker: PropTypes.bool,

  /** Display 24 hour military style time. Also hides am/pm selector. */
  timePicker24Hour: PropTypes.bool,

  /** Size of increments for minute selector. */
  timePickerIncrement: PropTypes.number,

  /** Show two timepickers to define a time range */
  timePickerRange: PropTypes.bool,

  /** Show seconds selector. timePicker or timePickerRange must be set */
  timePickerSeconds: PropTypes.bool,

  /** Display column left of calendar with week numbers */
  weekNumbers: PropTypes.bool,

  /** Show select dropdown for the Year in the calendar header */
  yearDropdown: PropTypes.bool,

  /** The maximum year in the select year dropdown */
  yearDropdownMaxYear: PropTypes.number,

  /** The minimum year in the select year dropdown */
  yearDropdownMinYear: PropTypes.number
};

DatePicker.defaultProps = {
  alwaysShowCalendars: false,
  applyButtonClasses: '',
  autoApply: false,
  buttonClasses: '',
  cancelButtonClasses: '',
  className: '',
  dataQa: '',
  dateEnd: undefined,
  dateMax: moment().add(50, 'y').format('YYYY-MM-DD'),
  dateMin: moment().add(-50, 'y').format('YYYY-MM-DD'),
  dateStart: undefined,
  error: '',
  format: 'D MMM YYYY',
  getClassNamesForDate: () => '',
  htmlId: '',
  isSelectableDate: () => false,
  isoWeekNumbers: false,
  labelText: '',
  linkedCalendars: true,
  monthDropdown: false,
  name: '',
  numberOfMonths: 1,
  onChangeCallback: () => {},
  placeholder: '',
  positionY: 'down',
  positionX: 'right',
  rangeDelimiter: 'â€“',
  ranges: false,
  rangePresets: {
    Today: [
      moment().format('YYYY-MM-DD'),
      moment().format('YYYY-MM-DD')
    ],
    Yesterday: [
      moment().subtract(1, 'day').format('YYYY-MM-DD'),
      moment().subtract(1, 'day').format('YYYY-MM-DD')
    ],
    'Last 7 Days': [
      moment().subtract(7, 'day').format('YYYY-MM-DD'),
      moment().format('YYYY-MM-DD')
    ],
    'Last 30 Days': [
      moment().subtract(30, 'day').format('YYYY-MM-DD'),
      moment().format('YYYY-MM-DD')
    ],
    'This Month': [
      moment().startOf('month').format('YYYY-MM-DD'),
      moment().format('YYYY-MM-DD')
    ],
    'Last Month': [
      moment().subtract(1, 'month').startOf('month').format('YYYY-MM-DD'),
      moment().subtract(1, 'month').endOf('month').format('YYYY-MM-DD')
    ]
  },
  screenReaderLabel: false,
  showCustomRangeLabel: false,
  singleDatePicker: false,
  timePicker: false,
  timePicker24Hour: false,
  timePickerIncrement: 1,
  timePickerRange: false,
  timePickerSeconds: false,
  weekNumbers: false,
  yearDropdown: false,
  yearDropdownMaxYear: parseInt(moment().add(50, 'y').format('YYYY'), 10),
  yearDropdownMinYear: parseInt(moment().add(-50, 'y').format('YYYY'), 10)
};

export const BareDatePicker = DatePicker;
export default withTheme(withStyles(styles)(DatePicker));
