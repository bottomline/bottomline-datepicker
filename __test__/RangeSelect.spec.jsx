import React from 'react';
import PropTypes from 'prop-types';
import { mount } from 'enzyme';

import { nextTheme, ThemeProvider } from '@glu/theming';
import DatePicker from '../src/DatePicker';
import '../src/themeDefaults';

const Wrapper = ({ children }) => (
  <ThemeProvider baseTheme={nextTheme}>
    {children}
  </ThemeProvider>
);

Wrapper.propTypes = {
  children: PropTypes.node.isRequired
};

describe('RangeSelect', () => {
  afterEach(() => {
    document.body.innerHTML = '';
  });

  it('should set range after clicking on a predefined range label', () => {
    const wrapper = mount(
      <Wrapper>
        <DatePicker
          htmlId="unique"
          {...{
            ranges: true,
            dateStart: '2018-06-26',
            dateEnd: '2018-06-27',
            rangePresets: {
              Christmas: [
                '2017-12-20',
                '2017-12-25'
              ]
            }
          }}
        />
      </Wrapper>
    );
    wrapper.find('input').simulate('focus');
    wrapper.find('[data-qa="range-select"] [data-qa="range-option-Christmas"]').simulate('click');
    const inputValue = wrapper.find('input').prop('value');
    expect(inputValue).toMatchSnapshot();
  });

  it('should show calendar when choosing ranges', () => {
    const wrapper = mount(
      <Wrapper>
        <DatePicker
          htmlId="unique"
          {...{
            ranges: true,
            alwaysShowCalendars: true
          }}
        />
      </Wrapper>
    );
    wrapper.find('input').simulate('focus');
    wrapper.find('[data-qa="range-select"] [data-qa="range-option-Today"]').simulate('click');
    const inputValue = wrapper.find('input').prop('value');
    expect(inputValue).toMatchSnapshot();
  });

  it('should show custom range label', () => {
    const wrapper = mount(
      <Wrapper>
        <DatePicker
          htmlId="unique"
          {...{
            ranges: true,
            showCustomRangeLabel: true
          }}
        />
      </Wrapper>
    );
    wrapper.find('input').simulate('focus');
    wrapper.find('[data-qa="range-select"] [data-qa="range-option-custom"]').simulate('click');
    const inputValue = wrapper.find('input').prop('value');
    expect(inputValue).toMatchSnapshot();
  });
});
