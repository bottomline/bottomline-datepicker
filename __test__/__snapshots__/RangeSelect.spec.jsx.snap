// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RangeSelect should set range after clicking on a predefined range label 1`] = `"20 Dec 2017â€“25 Dec 2017"`;

exports[`RangeSelect should show calendar when choosing ranges 1`] = `""`;

exports[`RangeSelect should show custom range label 1`] = `""`;
